### PR TITLE
[SPARK-32953][PYTHON][SQL] Add Arrow self_destruct support to toPandas

### PR DESCRIPTION
**Pull request by lidavidm**
### What changes were proposed in this pull request?

Creating a Pandas dataframe via Apache Arrow currently can use twice as much memory as the final result, because during the conversion, both Pandas and Arrow retain a copy of the data. Arrow has a "self-destruct" mode now (Arrow >= 0.16) to avoid this, by freeing each column after conversion. This PR integrates support for this in toPandas, handling a couple of edge cases: 

self_destruct has no effect unless the memory is allocated appropriately, which is handled in the Arrow serializer here. Essentially, the issue is that self_destruct frees memory column-wise, but Arrow record batches are oriented row-wise:

```
Record batch 0: allocation 0: column 0 chunk 0, column 1 chunk 0, ...
Record batch 1: allocation 1: column 0 chunk 1, column 1 chunk 1, ...
```

In this scenario, Arrow will drop references to all of column 0's chunks, but no memory will actually be freed, as the chunks were just slices of an underlying allocation. The PR copies each column into its own allocation so that memory is instead arranged as so:

```
Record batch 0: allocation 0 column 0 chunk 0, allocation 1 column 1 chunk 0, ...
Record batch 1: allocation 2 column 0 chunk 1, allocation 3 column 1 chunk 1, ...
```

The optimization is disabled by default, and can be enabled with the Spark SQL conf "spark.sql.execution.arrow.pyspark.selfDestruct.enabled" set to "true". We can't always apply this optimization because it's more likely to generate a dataframe with immutable buffers, which Pandas doesn't always handle well, and because it is slower overall (since it only converts one column at a time instead of in parallel).

### Why are the changes needed?

This lets us load larger datasets - in particular, with N bytes of memory, before we could never load a dataset bigger than N/2 bytes; now the overhead is more like N/1.25 or so.

### Does this PR introduce _any_ user-facing change?

Yes - it adds a new SQL conf "spark.sql.execution.arrow.pyspark.selfDestruct.enabled"

### How was this patch tested?

See the [mailing list](http://apache-spark-developers-list.1001551.n3.nabble.com/DISCUSS-Reducing-memory-usage-of-toPandas-with-Arrow-quot-self-destruct-quot-option-td30149.html) - it was tested with Python memory_profiler. Unit tests added to check memory within certain bounds and correctness with the option enabled.

### CODE REVIEW

File: python/pyspark/sql/pandas/conversion.py
Comment (HyukjinKwon) [inline python/pyspark/sql/pandas/conversion.py:117]: Yeah, I think it's better. In which case should we disable? Maybe we should enable it by default.

**Comment by HyukjinKwon** [comment]
cc @WeichenXu123 who already took a close look.

File: python/pyspark/sql/pandas/conversion.py
Comment (lidavidm) [reply python/pyspark/sql/pandas/conversion.py:117]: The worry is with running into things like https://github.com/pandas-dev/pandas/issues/35530 in which case the user may appreciate an escape hatch.

**Comment by lidavidm** [comment]
Following up here - any other comments? Does this look like a desirable feature, and how do we want to configure it?

File: python/pyspark/sql/pandas/conversion.py
Comment (HyukjinKwon) [inline python/pyspark/sql/pandas/conversion.py:37]: Let's name it as `selfDestruct` to make the naming role consistent.

File: python/pyspark/sql/pandas/conversion.py
Comment (HyukjinKwon) [inline python/pyspark/sql/pandas/conversion.py:113]: Would you mind leaving a comment on the codes about this set of parameter configurations?

File: python/pyspark/sql/pandas/serializers.py
Comment (HyukjinKwon) [inline python/pyspark/sql/pandas/serializers.py:95]: Sorry for asking a question without taking a close look but would you mind elaborating why we should do this?

**Comment by HyukjinKwon** [comment]
cc @BryanCutler FYI.

**Comment by HyukjinKwon** [comment]
ok to test

File: python/pyspark/sql/pandas/serializers.py
Comment (lidavidm) [reply python/pyspark/sql/pandas/serializers.py:102]: I added a comment, but please let me know if it's unclear.

File: python/pyspark/sql/pandas/conversion.py
Comment (lidavidm) [reply python/pyspark/sql/pandas/conversion.py:117]: Done.

File: python/pyspark/sql/pandas/conversion.py
Comment (lidavidm) [reply python/pyspark/sql/pandas/conversion.py]: Done.

File: python/pyspark/sql/pandas/conversion.py
Comment (BryanCutler) [inline python/pyspark/sql/pandas/conversion.py:37]: Would this be better as an sql config? Since this class is a mixin, I'm not sure the user would see this option from the public api?

File: python/pyspark/sql/pandas/conversion.py
Comment (BryanCutler) [inline python/pyspark/sql/pandas/conversion.py:106]: does this have any bearing on the buffers self destructing? is it taking into account how many reference counts there are before destructing?

File: python/pyspark/sql/pandas/conversion.py
Comment (BryanCutler) [reply python/pyspark/sql/pandas/conversion.py:121]: yeah, I think this option can lead to other side effects, best to disable by default I think.

File: python/pyspark/sql/pandas/serializers.py
Comment (BryanCutler) [reply python/pyspark/sql/pandas/serializers.py:102]: This looks a little strange to me too. Is `concat_arrays` actually doing anything here, and if so, wouldn't it do it for the case that `selfDestruct` is False too?

File: python/pyspark/sql/pandas/conversion.py
Comment (BryanCutler) [inline python/pyspark/sql/pandas/conversion.py:118]: Is this necessary to set with `self_destruct`? It might lead to Pandas doing more memory allocation later, I believe.

**Review by BryanCutler — COMMENTED** [review]
Thanks for doing this @lidavidm , I agree it could be useful in cases of working with large DataFrames. The last I saw was that the `self_destruct` option is experimental. Do you know if or when this might change? I'm a little unsure about adding experimental features, especially if it could lead to issues with the resulting Pandas DataFrame.

File: python/pyspark/sql/pandas/conversion.py
Comment (lidavidm) [reply python/pyspark/sql/pandas/conversion.py:37]: I'll change it.

File: python/pyspark/sql/pandas/conversion.py
Comment (lidavidm) [reply python/pyspark/sql/pandas/conversion.py:106]: Yes - we don't want to hold on to any other references to the buffers, we want the Table to be the only owner. I'll clarify this part here.

**Comment by lidavidm** [comment]
I can't reply to this inline somehow, but

> This looks a little strange to me too. Is concat_arrays actually doing anything here, and if so, wouldn't it do it for the case that selfDestruct is False too?

1) Yes, it's copying each column in the record batch into its own allocation, as explained in the latest commit. I couldn't find a more explicit way in Arrow to copy a column out of a record batch.
2) Yes, I can try to thread through a parameter, though I think it's harmless in this case.

File: python/pyspark/sql/pandas/conversion.py
Comment (lidavidm) [reply python/pyspark/sql/pandas/conversion.py:118]: Not quite _necessary_ but good to have, else we may end up allocating a large block if there are a lot of columns of the same type, defeating the point. Pandas may reconsolidate later but that's part of the issue of using Pandas.

**Comment by lidavidm** [comment]
Ah and for

> The last I saw was that the self_destruct option is experimental. Do you know if or when this might change? I'm a little unsure about adding experimental features, especially if it could lead to issues with the resulting Pandas DataFrame.

I can follow up on the experimental status, but AIUI, it's going to just be a long tail of "this Pandas operation didn't expect an immutable backing array" that we would need to flush out anyways over time. We can leave it turned off by default. Also, I see the PyArrow optimization in Spark is itself experimental anyways.

**Comment by lidavidm** [comment]
Running the demo again gives these two plots. While the memory usage looks identical, in the no-self-destruct case, Python gets OOMKilled, while it does not get OOMKilled in the other case. The reason why the memory usage looks so similar is that jemalloc doesn't immediately return unused memory to the OS, but rather, when under memory pressure, signals a background thread to start cleaning up memory; because of this delay, memory still gets consumed, but at a slower pace.

Without self-destruct

no-self-destruct

With self-destruct
self-destruct

```
import time

import pyarrow
pyarrow.jemalloc_set_decay_ms(0)

from pyspark.sql import SparkSession
from pyspark.sql.functions import rand

self_destruct = "true"
print('self_destruct:', self_destruct)
spark = SparkSession.builder \
    .master("local") \
    .appName("demo") \
    .config("spark.driver.maxResultSize", "8g") \
    .config("spark.driver.memory", "4g") \
    .config("spark.executor.memory", "512m") \
    .config("spark.worker.memory", "512m") \
    .config("spark.sql.execution.arrow.pyspark.enabled", "true") \
    .config("spark.sql.execution.arrow.pyspark.selfDestruct.enabled", "true") \
    .getOrCreate()

# 6 GiB dataframe. Tweak this to adjust for the amount of RAM you have
# (target > ~1/2 of free memory). I had ~8 GiB free for this demo.
# union() generates a dataframe that doesn't take so much memory in Java
rows = 2 ** 17
cols = 64
df = spark.range(0, rows).select(*[rand(seed=i) for i in range(cols)])
df = df.union(df).union(df).union(df).union(df).union(df)
df = df.union(df)
df = df.union(df)
df = df.union(df)

pdf = df.toPandas()

print('================ MEMORY USAGE:', sum(pdf.memory_usage()) / 2**20, "MiB")
# Give memory_profiler some more time
time.sleep(2)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1852]: Let's change it to 3.1.0.

Comment (lidavidm) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Done, thanks.

**Review by HyukjinKwon — COMMENTED** [review]
I am good with adding this feature but I think I should leave it to @BryanCutler who's also a committer in Arrow.

**Comment by HyukjinKwon** [comment]
@iidavidm, just out of curiosity, we can do it in pandas UDFs too, right? I can of course do it separately in another PR but I was wondering if I am understanding correctly.

**Comment by lidavidm** [comment]
@HyukjinKwon hmm, it should also work there, but I can confirm (I haven't looked at that codepath/can measure the memory usage there too). It will only work one way though (Spark -> Pandas), this optimization doesn't apply the other way (Pandas -> Spark), so maybe not as much of a benefit there.

**Comment by HyukjinKwon** [comment]
Ah, that's fine. That one, we can just do it separately. We might have to use self-destruct at `ArrowStreamPandasSerializer.arrow_to_pandas` to make it working but sure there wouldn't be so much benefits. Let's. do it separately anyway- should be off-topic. I was just curious :-).

**Comment by BryanCutler** [comment]
@lidavidm I'm a little swamped right now with some other things, but will take another look as soon as I can.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (BryanCutler) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1851]: Can you also add that the `split_blocks` will be enabled too, and make it clear that is is for the conversion of arrow to pandas?

File: python/pyspark/sql/pandas/serializers.py
Comment (BryanCutler) [inline python/pyspark/sql/pandas/serializers.py:58]: This doesn't need to be done in the serializer, could you move it to `_collect_as_arrow` and leave the serializer as it was?

File: python/pyspark/sql/pandas/serializers.py
Comment (BryanCutler) [inline python/pyspark/sql/pandas/serializers.py:67]: I'm not crazy about using `pa.concat_arrays()` to force a copy. I don't think it's safe to assume it will always make a copy, it might end up producing a chunked array in the future with zero copy, as is done in `pa.concat_tables()`. Is there another, more direct way we can make copies?

File: python/pyspark/sql/pandas/conversion.py
Comment (BryanCutler) [inline python/pyspark/sql/pandas/conversion.py:257]: you should be able to split the batches here, as they are yielded from the socket

File: python/pyspark/sql/tests/test_arrow.py
Comment (BryanCutler) [inline python/pyspark/sql/tests/test_arrow.py:195]: This is a pretty minimal dataframe, mostly for checking type compatibility. It would be nice to also check against a larger dataframe that the memory usage doesn't double during conversion, but I'm not sure how to do that reliably.

**Review by BryanCutler — CHANGES_REQUESTED** [review]
Sorry for the delay, @lidavidm . I'm not too crazy about having to reallocate all of the arrays so that they can be self-destructed. It would be good to have a `RecordBatchReader` with an option to allocate arrays in a way they can be freed later, but I'm not sure how feasible that is. Despite that, I think this is still a good option to minimize memory during Pandas conversion so it will be useful. I would like to see the array re-allocation done in `collect_as_arrow()` instead of in the serializer, though.

File: python/pyspark/sql/pandas/serializers.py
Comment (lidavidm) [reply python/pyspark/sql/pandas/serializers.py:67]: I'll look around. I couldn't find another obvious one but I agree that this is not ideal.

File: python/pyspark/sql/tests/test_arrow.py
Comment (lidavidm) [reply python/pyspark/sql/tests/test_arrow.py:195]: It's really hard to verify this in an automated way due to how jemalloc works (except possibly by reconfiguring jemalloc) - at the very least it'd be iffy to do in a unit test like this without some setup.

(The brief explanation is that with how Arrow configures jemalloc, memory is only slowly reclaimed by the OS, so in naive measurements, you'll still see a memory usage spike - albeit slightly less of one. In C++, you can force jemalloc to purge everything right away, but even via ctypes I couldn't access the right functions from Python.)

Comment (BryanCutler) [reply python/pyspark/sql/tests/test_arrow.py:195]: Well, I was thinking you might be able to use `pyarrow.total_allocated_bytes()` - this is how it's tested in pyarrow https://github.com/apache/arrow/blob/e6eb61f58ef382003c9462924563f575d9a59c13/python/pyarrow/tests/test_pandas.py#L3340. But the table is deallocated after `toPandas()` returns anyway, so you would have to sample before and after `pdf = table.to_pandas(**pandas_options)`..

Comment (lidavidm) [reply python/pyspark/sql/tests/test_arrow.py:195]: Hmm, is that test quite right? self_destruct doesn't change the end memory consumption, but rather the peak consumption...which you can't quite reliably measure from Python (via a separate thread, perhaps?).

File: python/pyspark/sql/pandas/serializers.py
Comment (lidavidm) [reply python/pyspark/sql/pandas/serializers.py:67]: I don't see an explicit way to copy. The take/filter kernels might suffice but would require constructing auxiliary arrays; we could also try copying buffers directly. Would it be better perhaps to defer this and work on a explicit copy in Arrow, or a RecordBatchReader that doesn't require copies?

Comment (BryanCutler) [reply python/pyspark/sql/pandas/serializers.py:67]: I don't think we need to defer since what you have works, but if you could look into those 2 improvements on the Arrow side that would be great. Then we could come back and update this later.

File: python/pyspark/sql/tests/test_arrow.py
Comment (BryanCutler) [reply python/pyspark/sql/tests/test_arrow.py:195]: The pyarrow test works because it allocates the memory for the required Pandas blocks too, which I believe doubles the memory allocated normally, but ends up with a net zero change with self destruct.

Comment (lidavidm) [reply python/pyspark/sql/tests/test_arrow.py]: Sorry again for the delay.

With some refactoring, this test now functions similarly to the PyArrow test. However if we don't want to rearrange things like this I can revert this change.

File: python/pyspark/sql/pandas/serializers.py
Comment (lidavidm) [reply python/pyspark/sql/pandas/serializers.py]: I filed https://issues.apache.org/jira/browse/ARROW-10670.

File: python/pyspark/sql/tests/test_arrow.py
Comment (BryanCutler) [inline python/pyspark/sql/tests/test_arrow.py:208]: just to be safe, what do you think about getting allocated bytes before and after, then comparing difference? It should probably be the same, but would be a little more focused then in case something changes in the future

File: python/pyspark/sql/tests/test_arrow.py
Comment (BryanCutler) [inline python/pyspark/sql/tests/test_arrow.py:215]: looks to be unused?

File: python/pyspark/sql/pandas/conversion.py
Comment (BryanCutler) [inline python/pyspark/sql/pandas/conversion.py:245]: I don't think we really need to add this method, it just does the conversion from batches to DataFrame. The unit test doesn't even use the returned Table right?

**Review by BryanCutler — CHANGES_REQUESTED** [review]
It doesn't seem to me we really need this refactoring. Will your unit test still pass if you get only call `toPandas()`?

File: python/pyspark/sql/tests/test_arrow.py
Comment (BryanCutler) [inline python/pyspark/sql/tests/test_arrow.py:203]: Oh, I missed this comment that you need the Table reference. Otherwise it could be freed and then we lose track of memory allocated right? If that's the case, what about just calling `_collect_as_arrow(_force_split_batches=True)` here and perform batches -> table -> pandas directly?

Comment (lidavidm) [reply python/pyspark/sql/tests/test_arrow.py:203]: Yes - we need the table reference. If self destruct worked, then it holds effectively 0 memory; if it didn't work, then it holds a copy of the dataframe. 

If we call to_pandas inside the test itself, wouldn't we then be testing Arrow directly, instead of testing how Arrow is used inside Spark?

File: python/pyspark/sql/tests/test_arrow.py
Comment (lidavidm) [reply python/pyspark/sql/tests/test_arrow.py:208]: Hmm, I think that would work if we compare it after we delete the dataframe (but not the table). Then self_destruct should ensure that Arrow doesn't hold any copies of data.

File: python/pyspark/sql/tests/test_arrow.py
Comment (BryanCutler) [reply python/pyspark/sql/tests/test_arrow.py:203]: Not really, it's not ideal but it would still test that splitting and reallocating the columns in `_collect_as_arrow()` will allow for self destruction, which is most of additions here. The rest is just calling `to_pandas()` with the right options. I would recommend testing something like this:

```python
with self.sql_conf({"spark.sql.execution.arrow.pyspark.selfDestruct.enabled": True}):
    # get initial pa.total_allocated_bytes()
    batches = _collect_as_arrow(_force_split_batches=True)
    table = # convert batches to table as done in toPandas()
    pdf = table.to_pandas(self_destruct=True, split_blocks=True, use_threads=False)
    # get after pa.total_allocated_bytes() and check difference is expected

    # Call with the full code path and compare resulting DataFrame for good measure
    assert_frames_equal(df.toPandas(), pdf)
```

That should be sufficient testing, wdyt?

Comment (lidavidm) [reply python/pyspark/sql/tests/test_arrow.py:203]: Makes sense to me - I'll try to update this soon. Thanks for being patient with me here.

Also looks like the 3.1 branch was cut already, so I need to update the flags here to reference 3.2.

**Comment by lidavidm** [comment]
Hmm, I'm not sure what's with javadoc generation failing on all the CI tests...

File: python/pyspark/sql/tests/test_arrow.py
Comment (BryanCutler) [inline python/pyspark/sql/tests/test_arrow.py:214]: This block isn't quite what I was suggesting. We should compare the Pandas DataFrame resulting from `df.toPandas()` with `spark.sql.execution.arrow.pyspark.selfDestruct.enabled` False to `pdf_split`. Also, we will need to call `df.toPandas()` with `spark.sql.execution.arrow.pyspark.selfDestruct.enabled` True and compare that as well.

File: python/pyspark/sql/pandas/conversion.py
Comment (BryanCutler) [inline python/pyspark/sql/pandas/conversion.py:284]:
```suggestion
                    if isinstance(batch_or_indices, pa.RecordBatch):
                        batch_or_indices = pa.RecordBatch.from_arrays([
                            pa.concat_arrays([array])
                            for array in batch_or_indices
                        ], schema=batch_or_indices.schema)
                    results.append(batch_or_indices)
```

**Review by BryanCutler — CHANGES_REQUESTED** [review]
Sorry for the delay @lidavidm . I think the code changes look good now, just a few issues with the testing remain. We need to cover:

1. Test that memory stays within a certain bound (which you have)
2. Test complete code path with the new config enabled
3. Verify the 2 results above are equal to result of `df.toPandas()` with config disabled

File: python/pyspark/sql/tests/test_arrow.py
Comment (BryanCutler) [reply python/pyspark/sql/tests/test_arrow.py:196]: how about  `2 ** 10` ?

**Review by BryanCutler — APPROVED** [review]
LGTM, I'm going to run tests once more for good measure and I'll merge after passing.

**Comment by BryanCutler** [comment]
restest this please

**Comment by lidavidm** [comment]
Thanks for being patient with me through this very long review!

**Comment by BryanCutler** [comment]
Merged to master, apologies for all of the delays and thanks @lidavidm for all the work! We should add this to the documentation too, if you could do that it would be great.

**Pull request closed without merging**
